src/plans/D3MAavePlan.sol

```diff
@@ -0,0 +1,209 @@
+    InterestRateStrategyLike     public tack;
+    uint256                      public bar;  // Target Interest Rate [ray]
+
+    LendingPoolLike public immutable pool;
+    TokenLike       public immutable stableDebt;
+    TokenLike       public immutable variableDebt;
+    TokenLike       public immutable dai;
+    address         public immutable adai;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed what, address data);
+
+    constructor(address dai_, address pool_) {
+        dai = TokenLike(dai_);
+        pool = LendingPoolLike(pool_);
+
+        // Fetch the reserve data from Aave
+        (,,,,,,, address adai_, address stableDebt_, address variableDebt_, address interestStrategy_,) = LendingPoolLike(pool_).getReserveData(dai_);
+        require(adai_ != address(0), "D3MAavePlan/invalid-adai");
+        require(stableDebt_ != address(0), "D3MAavePlan/invalid-stableDebt");
+        require(variableDebt_ != address(0), "D3MAavePlan/invalid-variableDebt");
+        require(interestStrategy_ != address(0), "D3MAavePlan/invalid-interestStrategy");
+
+        adai = adai_;
+        stableDebt = TokenLike(stableDebt_);
+        variableDebt = TokenLike(variableDebt_);
+        tack = InterestRateStrategyLike(interestStrategy_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "D3MAavePlan/not-authorized");
+        _;
+    }
+
+    // --- Math ---
+    uint256 constant RAY  = 10 ** 27;
+    function _rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x * y) / RAY;
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x * RAY) / y;
+    }
+
+    // --- Admin ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "bar") {
+            bar = data;
+        } else revert("D3MAavePlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "tack") tack = InterestRateStrategyLike(data);
+        else revert("D3MAavePlan/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function _maxBar() internal view returns (uint256) {
+        return tack.getMaxVariableBorrowRate();
+    }
+
+    // --- Automated Rate targeting ---
+    function _calculateTargetSupply(uint256 targetInterestRate, uint256 totalDebt) internal view returns (uint256) {
+        uint256 base = tack.baseVariableBorrowRate();
+        if (targetInterestRate <= base || targetInterestRate > _maxBar()) {
```

Comment:
@kmbarry1 suggested that if `targetInterestRate <= base` we will wind as much as possible (https://github.com/makerdao/dss-direct-deposit/pull/51#discussion_r891486730).
I think that is reasonable for `targetInterestRate -= base` as a means for gov to easily wind as much as possible, and in that case we can return `type(uint256).max`.
Still think that for `targetInterestRate < base` we should return 0, as that is an illegal value (for both aave and compound base is 0 atm anyway).
@iamchrissmith @hexonaut @kmbarry1 let me know what you think.

Reply:
This line evolved from a require that exists in the current Aave D3M (https://github.com/makerdao/dss-direct-deposit/blob/master/src/DssDirectDepositAaveDai.sol#L230).  I agree with @kmbarry1 's logic, but would want to make sure from @hexonaut that there isn't a weird thing with Aave that `== base` isn't somehow invalid

Reply:
This is definitely an illegal value imo. I get the sentiment, but my preference is to turn the thing off if we have supplied an invalid value vs try to max it out.

Reply:
@hexonaut just to be clear, @kmbarry1 and @talbaneth 's comment about about the case where our `bar == base` so not the max, but the min. 

But in looking into their current implementation the `base == 0` so if our target is 0 we would already have returned at this point.  If we were to remove `if (targetInterestRate == 0) return 0;  // De-activated` in `getTargetAssets` for some reason AND changed this check to `<` then we could set `bar = 0` and end up with a div zero revert. 

However, I think the real question is, if they change their base to something `> 0`, so we want to proceed with the calculation or do we want to unwind our position because `bar == base`?

It looks like from the aave comment that the base represents their borrow rate when utilization is 0, so if we were to decide `bar == base` then we are essentially saying we only want to supply when no one is borrowing, I think

Reply:
According to theory `targetInterestRate == base` is not an invalid edge case (see [section 2.5](https://github.com/aave/aave-protocol/blob/master/docs/Aave_Protocol_Whitepaper_v1_0.pdf)). I think that's enough to be able to use it for gov to indicate full wind if we want.

Codewise I don't see a fundamental problem with changing the condition in aave plan to:
`if (targetInterestRate < base..`
and adding:
`if (targetInterestRate == base) return uint256(type(uint256).max);

It will also be required to change the indication for an inacitve plan to something else than `bar == 0` (which might actually be more elegant and fitting `disabled`/`paused`/`live` flags that we usually use elsewehre).

Not sure if we want to do these changes at this point as we want/ed to freeze the code.
This is not a bug or security issue (more of governance convenience/simplicity thing) so inclined not to do it unless there are further delays. Let me know what you guys think.

Reply:
i'm not sure i understand or agree with `if (targetInterestRate == base) return uint256(type(uint256).max);` why would we return max uint for target assets if we were at the base?

Reply:
but i do agree that `targetInterestRate == base` is valid.

Reply:
Regarding why would we return max uint for target assets if we were at the base - that is the lowest viable rate, and we always inject more dai to lower the rate. If we really want base rate than it would only happen when there are no borrows and it would not change how much dai we inject in that situation, but the idea is to use it as a rate that gov can set to indicate full winding.
See Kurt's initial comment - (https://github.com/makerdao/dss-direct-deposit/pull/51#discussion_r891486730)

Regarding why it would need to change the bar == 0 for being not active - base is currently 0 in both aave and compound, so if we use bar == 0 as a means for gov to indicate full winding we need some other internal indication for full unwinding (which is currently bar == 0).

Anyway, I tend to think not to do this change atm, but Since Kurt brought it up and it makes sense we opened it for dicussion.


bar == 0

Reply:
thank you @talbaneth those explanations help/make sense.  

I am also leaning now towards not changing this. If we see that Aave changes their base value we can re-evaluate i think

Reply:
This was discussed again this week and deemed ok not to support governance indicating wind with a zero rate.